Data race on `cert_cb` (openssl)

### Describe the bug

https://s3.amazonaws.com/clickhouse-test-reports/PRs/85378/9c3c95dd94efdde041befc60b845404f6a329b82//integration_tests_tsan_4_6/integration_run_parallel_0.log

### How to reproduce

_No response_

### Error message and/or stacktrace

_No response_

```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Read of size 8 at 0x72300000e8e0 by thread T601 (mutexes: write M0, write M1):
E               #0 ssl_cert_dup ci/tmp/build/./contrib/openssl/ssl/ssl_cert.c:197:26 (clickhouse+0x28d2c3b3) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 ossl_ssl_connection_new_int ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:775:15 (clickhouse+0x28d352af) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 ossl_ssl_connection_new ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:930:12 (clickhouse+0x28d35fe8) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #3 SSL_new ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:692:12 (clickhouse+0x28d34db9) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 Poco::Net::SecureSocketImpl::acceptSSL() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:110:10 (clickhouse+0x2ad06dcc) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 Poco::Net::SecureStreamSocketImpl::acceptSSL() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureStreamSocketImpl.cpp:86:8 (clickhouse+0x2ad0de9d) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 Poco::Net::SecureSocketImpl::acceptConnection(Poco::Net::SocketAddress&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:95:27 (clickhouse+0x2ad06b76) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 Poco::Net::SecureServerSocketImpl::acceptConnection(Poco::Net::SocketAddress&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureServerSocketImpl.cpp:43:15 (clickhouse+0x2ad057e4) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #8 Poco::Net::ServerSocket::acceptConnection() ci/tmp/build/./base/poco/Net/src/ServerSocket.cpp:154:30 (clickhouse+0x2acd62c4) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #9 Poco::Net::TCPServer::run() ci/tmp/build/./base/poco/Net/src/TCPServer.cpp:137:47 (clickhouse+0x2ace8b5b) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #10 Poco::(anonymous namespace)::RunnableHolder::run() ci/tmp/build/./base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x2ac56baf) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #11 Poco::ThreadImpl::runnableEntry(void*) ci/tmp/build/./base/poco/Foundation/src/Thread_POSIX.cpp:341:27 (clickhouse+0x2ac54e09) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Previous write of size 8 at 0x72300000e8e0 by main thread (mutexes: write M2):
E               #0 ssl_cert_set_cert_cb ci/tmp/build/./contrib/openssl/ssl/ssl_cert.c:406:16 (clickhouse+0x28d2cfa4) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 SSL_CTX_set_cert_cb ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:4314:5 (clickhouse+0x28d3da39) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 DB::CertificateReloader::init(DB::CertificateReloader::MultiData*) ci/tmp/build/./src/Server/CertificateReloader.cpp:90:5 (clickhouse+0x21afffdf) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #3 DB::CertificateReloader::tryLoadImpl(Poco::Util::AbstractConfiguration const&, ssl_ctx_st*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) ci/tmp/build/./src/Server/CertificateReloader.cpp:170:17 (clickhouse+0x21b01c1c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 DB::CertificateReloader::tryLoad(Poco::Util::AbstractConfiguration const&, ssl_ctx_st*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) ci/tmp/build/./src/Server/CertificateReloader.cpp:110:5 (clickhouse+0x21b003b8) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 DB::CertificateReloader::tryLoad(Poco::Util::AbstractConfiguration const&) ci/tmp/build/./src/Server/CertificateReloader.cpp:97:5 (clickhouse+0x21b003b8)
E               #6 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) ci/tmp/build/./programs/server/Server.cpp:2638:41 (clickhouse+0x1333b4da) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x2ad211be) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #8 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:622:25 (clickhouse+0x1331e9d6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #9 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x2ad3ed60) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #10 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:409:20 (clickhouse+0x1331b54c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #11 main ci/tmp/build/./programs/main.cpp:381:21 (clickhouse+0x9026b84) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Location is heap block of size 192 at 0x72300000e880 allocated by main thread:
E               #0 malloc <null> (clickhouse+0x8fa09a0) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 CRYPTO_malloc ci/tmp/build/./contrib/openssl/crypto/mem.c:202:11 (clickhouse+0x28eee514) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 CRYPTO_zalloc ci/tmp/build/./contrib/openssl/crypto/mem.c:222:11 (clickhouse+0x28eee514)
E               #3 ssl_cert_new ci/tmp/build/./contrib/openssl/ssl/ssl_cert.c:69:11 (clickhouse+0x28d2ba7f) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 SSL_CTX_new_ex ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:3955:22 (clickhouse+0x28d3d04b) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 SSL_CTX_new ci/tmp/build/./contrib/openssl/ssl/ssl_lib.c:4114:12 (clickhouse+0x28d3d75d) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 Poco::Net::Context::createSSLContext() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/Context.cpp (clickhouse+0x2aced8fc) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 Poco::Net::Context::init(Poco::Net::Context::Params const&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/Context.cpp:234:2 (clickhouse+0x2acebfae) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #8 Poco::Net::Context::Context(Poco::Net::Context::Usage, Poco::Net::Context::Params const&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/Context.cpp:54:2 (clickhouse+0x2acebf12) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #9 Poco::Net::SSLManager::initDefaultContext(bool) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SSLManager.cpp:284:34 (clickhouse+0x2acf610f) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #10 Poco::Net::SSLManager::defaultServerContext() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SSLManager.cpp:125:3 (clickhouse+0x2acf4aa2) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #11 Poco::Net::SecureServerSocket::SecureServerSocket() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureServerSocket.cpp:30:65 (clickhouse+0x2ad044f8) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #12 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11::operator()(unsigned short) const ci/tmp/build/./programs/server/Server.cpp:2272:51 (clickhouse+0x133635cb) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #13 decltype(std::declval<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&>()(std::declval<unsigned short>())) std::__1::__invoke[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:149:25 (clickhouse+0x133635cb)
E               #14 DB::ProtocolServerAdapter std::__1::__invoke_void_return_wrapper<DB::ProtocolServerAdapter, false>::__call[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:216:12 (clickhouse+0x133635cb)
E               #15 std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11, DB::ProtocolServerAdapter (unsigned short)>::operator()[abi:ne190107](unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:210:12 (clickhouse+0x133635cb)
E               #16 DB::ProtocolServerAdapter std::__1::__function::__policy_invoker<DB::ProtocolServerAdapter (unsigned short)>::__call_impl[abi:ne190107]<std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11, DB::ProtocolServerAdapter (unsigned short)>>(std::__1::__function::__policy_storage const*, unsigned short) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:610:12 (clickhouse+0x133635cb)
E               #17 std::__1::__function::__policy_func<DB::ProtocolServerAdapter (unsigned short)>::operator()[abi:ne190107](unsigned short&&) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:716:12 (clickhouse+0x1331d433) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #18 std::__1::function<DB::ProtocolServerAdapter (unsigned short)>::operator()(unsigned short) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:989:10 (clickhouse+0x1331d433)
E               #19 DB::Server::createServer(Poco::Util::AbstractConfiguration&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, char const*, bool, bool, std::__1::vector<DB::ProtocolServerAdapter, std::__1::allocator<DB::ProtocolServerAdapter>>&, std::__1::function<DB::ProtocolServerAdapter (unsigned short)>&&) const ci/tmp/build/./programs/server/Server.cpp:546:27 (clickhouse+0x1331d433)
E               #20 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) ci/tmp/build/./programs/server/Server.cpp:2266:13 (clickhouse+0x13336b8e) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #21 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x2ad211be) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #22 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:622:25 (clickhouse+0x1331e9d6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #23 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x2ad3ed60) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #24 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:409:20 (clickhouse+0x1331b54c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #25 main ci/tmp/build/./programs/main.cpp:381:21 (clickhouse+0x9026b84) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Mutex M0 (0x723400004cc8) created at:
E               #0 pthread_mutex_init <null> (clickhouse+0x8fa3e93) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 std::__1::__libcpp_recursive_mutex_init[abi:ne190107](pthread_mutex_t*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__thread/support/pthread.h:62:10 (clickhouse+0x2bdef884) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 std::__1::recursive_mutex::recursive_mutex() ci/tmp/build/./contrib/llvm-project/libcxx/src/mutex.cpp:46:12 (clickhouse+0x2bdef884)
E               #3 Poco::Net::SecureSocketImpl::SecureSocketImpl(Poco::AutoPtr<Poco::Net::SocketImpl>, Poco::AutoPtr<Poco::Net::Context>) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:63:19 (clickhouse+0x2ad06507) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 Poco::Net::SecureServerSocketImpl::SecureServerSocketImpl(Poco::AutoPtr<Poco::Net::Context>) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureServerSocketImpl.cpp:23:2 (clickhouse+0x2ad055af) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 Poco::Net::SecureServerSocket::SecureServerSocket() ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureServerSocket.cpp:30:19 (clickhouse+0x2ad04507) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11::operator()(unsigned short) const ci/tmp/build/./programs/server/Server.cpp:2272:51 (clickhouse+0x133635cb) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 decltype(std::declval<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&>()(std::declval<unsigned short>())) std::__1::__invoke[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:149:25 (clickhouse+0x133635cb)
E               #8 DB::ProtocolServerAdapter std::__1::__invoke_void_return_wrapper<DB::ProtocolServerAdapter, false>::__call[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11&, unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:216:12 (clickhouse+0x133635cb)
E               #9 std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11, DB::ProtocolServerAdapter (unsigned short)>::operator()[abi:ne190107](unsigned short&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:210:12 (clickhouse+0x133635cb)
E               #10 DB::ProtocolServerAdapter std::__1::__function::__policy_invoker<DB::ProtocolServerAdapter (unsigned short)>::__call_impl[abi:ne190107]<std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_11, DB::ProtocolServerAdapter (unsigned short)>>(std::__1::__function::__policy_storage const*, unsigned short) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:610:12 (clickhouse+0x133635cb)
E               #11 std::__1::__function::__policy_func<DB::ProtocolServerAdapter (unsigned short)>::operator()[abi:ne190107](unsigned short&&) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:716:12 (clickhouse+0x1331d433) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #12 std::__1::function<DB::ProtocolServerAdapter (unsigned short)>::operator()(unsigned short) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:989:10 (clickhouse+0x1331d433)
E               #13 DB::Server::createServer(Poco::Util::AbstractConfiguration&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, char const*, bool, bool, std::__1::vector<DB::ProtocolServerAdapter, std::__1::allocator<DB::ProtocolServerAdapter>>&, std::__1::function<DB::ProtocolServerAdapter (unsigned short)>&&) const ci/tmp/build/./programs/server/Server.cpp:546:27 (clickhouse+0x1331d433)
E               #14 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) ci/tmp/build/./programs/server/Server.cpp:2266:13 (clickhouse+0x13336b8e) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #15 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x2ad211be) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #16 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:622:25 (clickhouse+0x1331e9d6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #17 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x2ad3ed60) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #18 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:409:20 (clickhouse+0x1331b54c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #19 main ci/tmp/build/./programs/main.cpp:381:21 (clickhouse+0x9026b84) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Mutex M1 (0x723800078030) created at:
E               #0 pthread_mutex_init <null> (clickhouse+0x8fa3e93) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 std::__1::__libcpp_recursive_mutex_init[abi:ne190107](pthread_mutex_t*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__thread/support/pthread.h:62:10 (clickhouse+0x2bdef884) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 std::__1::recursive_mutex::recursive_mutex() ci/tmp/build/./contrib/llvm-project/libcxx/src/mutex.cpp:46:12 (clickhouse+0x2bdef884)
E               #3 Poco::Net::SecureSocketImpl::SecureSocketImpl(Poco::AutoPtr<Poco::Net::SocketImpl>, Poco::AutoPtr<Poco::Net::Context>) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:63:19 (clickhouse+0x2ad06507) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 Poco::Net::SecureStreamSocketImpl::SecureStreamSocketImpl(Poco::Net::StreamSocketImpl*, Poco::AutoPtr<Poco::Net::Context>) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureStreamSocketImpl.cpp:34:2 (clickhouse+0x2ad0d6e3) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 Poco::Net::SecureSocketImpl::acceptConnection(Poco::Net::SocketAddress&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureSocketImpl.cpp:94:70 (clickhouse+0x2ad06b29) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 Poco::Net::SecureServerSocketImpl::acceptConnection(Poco::Net::SocketAddress&) ci/tmp/build/./base/poco/NetSSL_OpenSSL/src/SecureServerSocketImpl.cpp:43:15 (clickhouse+0x2ad057e4) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 Poco::Net::ServerSocket::acceptConnection() ci/tmp/build/./base/poco/Net/src/ServerSocket.cpp:154:30 (clickhouse+0x2acd62c4) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #8 Poco::Net::TCPServer::run() ci/tmp/build/./base/poco/Net/src/TCPServer.cpp:137:47 (clickhouse+0x2ace8b5b) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #9 Poco::(anonymous namespace)::RunnableHolder::run() ci/tmp/build/./base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x2ac56baf) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #10 Poco::ThreadImpl::runnableEntry(void*) ci/tmp/build/./base/poco/Foundation/src/Thread_POSIX.cpp:341:27 (clickhouse+0x2ac54e09) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Mutex M2 (0x555a7f7f83e0) created at:
E               #0 pthread_mutex_lock <null> (clickhouse+0x8fa41ce) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 std::__1::__libcpp_mutex_lock[abi:ne190107](pthread_mutex_t*) ci/tmp/build/./contrib/llvm-project/libcxx/include/__thread/support/pthread.h:95:10 (clickhouse+0x2bdef759) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 std::__1::mutex::lock() ci/tmp/build/./contrib/llvm-project/libcxx/src/mutex.cpp:29:12 (clickhouse+0x2bdef759)
E               #3 std::__1::lock_guard<std::__1::mutex>::lock_guard[abi:ne190107](std::__1::mutex&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__mutex/lock_guard.h:33:10 (clickhouse+0x21b03d29) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 DB::CertificateReloader::tryReloadAll(Poco::Util::AbstractConfiguration const&) ci/tmp/build/./src/Server/CertificateReloader.cpp:182:21 (clickhouse+0x21b03d29)
E               #5 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0::operator()(Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool) const ci/tmp/build/./programs/server/Server.cpp:2178:45 (clickhouse+0x1335b934) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 decltype(std::declval<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&>()(std::declval<Poco::AutoPtr<Poco::Util::AbstractConfiguration>>(), std::declval<bool>())) std::__1::__invoke[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:149:25 (clickhouse+0x133572a9) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #7 void std::__1::__invoke_void_return_wrapper<void, true>::__call[abi:ne190107]<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool>(DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__type_traits/invoke.h:224:5 (clickhouse+0x133572a9)
E               #8 std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0, void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::operator()[abi:ne190107](Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:210:12 (clickhouse+0x133572a9)
E               #9 void std::__1::__function::__policy_invoker<void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::__call_impl[abi:ne190107]<std::__1::__function::__default_alloc_func<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0, void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>>(std::__1::__function::__policy_storage const*, Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool) ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:610:12 (clickhouse+0x133572a9)
E               #10 std::__1::__function::__policy_func<void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::operator()[abi:ne190107](Poco::AutoPtr<Poco::Util::AbstractConfiguration>&&, bool&&) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:716:12 (clickhouse+0x22a62309) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #11 std::__1::function<void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>::operator()(Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool) const ci/tmp/build/./contrib/llvm-project/libcxx/include/__functional/function.h:989:10 (clickhouse+0x22a62309)
E               #12 DB::ConfigReloader::reloadIfNewer(bool, bool, bool, bool) ci/tmp/build/./src/Common/Config/ConfigReloader.cpp:179:13 (clickhouse+0x22a62309)
E               #13 DB::ConfigReloader::ConfigReloader(std::__1::basic_string_view<char, std::__1::char_traits<char>>, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, zkutil::ZooKeeperNodeCache&&, std::__1::shared_ptr<Poco::Event> const&, std::__1::function<void (Poco::AutoPtr<Poco::Util::AbstractConfiguration>, bool)>&&) ci/tmp/build/./src/Common/Config/ConfigReloader.cpp:38:19 (clickhouse+0x22a607a6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #14 std::__1::__unique_if<DB::ConfigReloader>::__unique_single std::__1::make_unique[abi:ne190107]<DB::ConfigReloader, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, zkutil::ZooKeeperNodeCache, std::__1::shared_ptr<Poco::Event>&, DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>>&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, zkutil::ZooKeeperNodeCache&&, std::__1::shared_ptr<Poco::Event>&, DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_0&&) ci/tmp/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:634:30 (clickhouse+0x13336539) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #15 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) ci/tmp/build/./programs/server/Server.cpp:1906:33 (clickhouse+0x13336539)
E               #16 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x2ad211be) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #17 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:622:25 (clickhouse+0x1331e9d6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #18 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x2ad3ed60) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #19 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:409:20 (clickhouse+0x1331b54c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #20 main ci/tmp/build/./programs/main.cpp:381:21 (clickhouse+0x9026b84) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E             Thread T601 (tid=609, running) created by main thread at:
E               #0 pthread_create <null> (clickhouse+0x8fa24c1) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #1 Poco::ThreadImpl::startImpl(Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>) ci/tmp/build/./base/poco/Foundation/src/Thread_POSIX.cpp:203:7 (clickhouse+0x2ac546c7) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #2 Poco::Thread::start(Poco::Runnable&) ci/tmp/build/./base/poco/Foundation/src/Thread.cpp:118:2 (clickhouse+0x2ac564af) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #3 Poco::Net::TCPServer::start() ci/tmp/build/./base/poco/Net/src/TCPServer.cpp:111:13 (clickhouse+0x2ace89e6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #4 DB::ProtocolServerAdapter::TCPServerAdapterImpl::start() ci/tmp/build/./src/Server/ProtocolServerAdapter.cpp:17:41 (clickhouse+0x21c22ee3) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #5 DB::ProtocolServerAdapter::start() ci/tmp/build/./src/Server/ProtocolServerAdapter.h:41:26 (clickhouse+0x13336fe8) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #6 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) ci/tmp/build/./programs/server/Server.cpp:2347:20 (clickhouse+0x13336fe8)
E               #7 Poco::Util::Application::run() ci/tmp/build/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x2ad211be) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #8 DB::Server::run() ci/tmp/build/./programs/server/Server.cpp:622:25 (clickhouse+0x1331e9d6) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #9 Poco::Util::ServerApplication::run(int, char**) ci/tmp/build/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x2ad3ed60) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #10 mainEntryClickHouseServer(int, char**) ci/tmp/build/./programs/server/Server.cpp:409:20 (clickhouse+0x1331b54c) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E               #11 main ci/tmp/build/./programs/main.cpp:381:21 (clickhouse+0x9026b84) (BuildId: d92946f55650e4d5914317a110eae52875789f19)
E           
E           SUMMARY: ThreadSanitizer: data race ci/tmp/build/./contrib/openssl/ssl/ssl_cert.c:197:26 in ssl_cert_dup
E           ==================
```

--- a/programs/server/Server.cpp
+++ b/programs/server/Server.cpp
@@ -2343,6 +2343,14 @@ try
     }
 
 #if USE_SSL
+    /// We may notice that try to reload certificates twice within this function.
+    /// First time here before starting the `servers_to_start_before_tables`, and then
+    /// one more time before starting the `servers`. The problem we're trying to avoid is:
+    /// 1. We create a socket for keeper secure protocol (w/o starting a server yet). That creates a SSL context (in particular, it might be the default context).
+    /// 2. We then start this server a few lines below. While handling the first request, it will access (read) certs.
+    /// 3. A little further down, we reload certificates before starting the rest of the servers.
+    /// 4. `CertificateReloader` will set its custom `cert_cb` to the default context if it is not initialized yet (see `CertificateReloader::init()`).
+    /// 5. Items (2) and (4) are not synchronized, so there might be a data race for example (see https://github.com/ClickHouse/ClickHouse/issues/85412).
     CertificateReloader::instance().tryLoad(config());
     CertificateReloader::instance().tryLoadClient(config());
 #endif

--- a/programs/server/Server.cpp
+++ b/programs/server/Server.cpp
@@ -2340,8 +2340,15 @@ try
             async_metrics,
             servers_to_start_before_tables,
             /* start_servers= */ false);
+    }
 
+#if USE_SSL
+    CertificateReloader::instance().tryLoad(config());
+    CertificateReloader::instance().tryLoadClient(config());
+#endif
 
+    {
+        std::lock_guard lock(servers_lock);
         for (auto & server : servers_to_start_before_tables)
         {
             server.start();

